tests/test_utils/__init__.py

```diff
@@ -1,36 +1,82 @@
-
-        value1 = dict1[key]
-        value2 = dict2[key]
-
-        if isinstance(value1, dict) and isinstance(value2, dict):
-            if not compare_dicts(value1, value2):
-                return False
-        elif isinstance(value1, list) and isinstance(value2, list):
-            if len(value1) != len(value2):
-                return False
-            # check if the lists contain dictionaries as elements
-            if len(value1) > 0 and isinstance(value1[0], dict) and isinstance(value2[0], dict):
-                for i in range(len(value1)):
-                    if not compare_dicts(value1[i], value2[i]):
-                        return False
-            elif sorted(value1) != sorted(value2):
-                return False
-        else:
-            if value1 != value2:
-                return False
-
-    return True
+import contextlib
+from contextlib import contextmanager
+
+
+class CompareDictsAddin:
+    def assert_dicts_equal(self, dict1, dict2, fail_fast=True, max_diff=None):
+        """
+        This function recursive compares two dictionaries handling cases where the values
+        are unordered arrays with elements that could be dictionaries.
+        :param dict1: first dictionary
+        :param dict2: second dictionary
+        :param fail_fast: if True, raise an exception on the first mismatch
+        :param max_diff: maximum number of differences to report
+        """
+        _mismatch = []
+        _path = ["$"]
+
+        class Stop(Exception):
+            pass
+
+        def update_mismatch(message):
+            _mismatch.append(message)
+            if fail_fast:
+                raise Stop()
+
+        @contextmanager
+        def path(key):
+            try:
+                yield _path.append(key)
+            finally:
+                _path.pop()
+
+        def _compare_dict(_dict1, _dict2):
+            if len(_dict1) != len(_dict2):
+                update_mismatch("{}: Lengths differ: {} vs {}".format(".".join(path), len(_dict1), len(_dict2)))
+
+            for key in dict1:
+                if key not in _dict2:
+                    update_mismatch("{}: Key {} not in dict2".format(".".join(path), key))
+                    continue
+                value1 = _dict1[key]
+                value2 = _dict2[key]
+
+                with path(key):
+                    if isinstance(value1, dict) and isinstance(value2, dict):
+                        _compare_dict(value1, value2)
+                    elif isinstance(value1, list) and isinstance(value2, list):
+                        if len(value1) != len(value2):
+                            update_mismatch(
+                                "{}: Lengths differ: {} vs {}".format(".".join(path), len(value1), len(value2))
+                            )
+                        # check if the lists contain dictionaries as elements
+                        elif len(value1) > 0 and isinstance(value1[0], dict) and isinstance(value2[0], dict):
```

Comment:
i think it's technically possible for a list to contain a dictionary and other things right? so you could have
```
list_of_things = [{"key": "value"}, 3, "string"]
```
so i think you'd have to iterate through all values in the list, and run `_compare_dict` for each value if they're different :|